Aprimoramento na busca de discipinas

Samlibardi opened the pull request:
Alterei o algoritmo de busca de disciplinas com base na busca por trigramas ([fonte](https://ii.nlm.nih.gov/MTI/Details/trigram.shtml)) de maneira a:

- encontrar disciplinas por meio de código ou nome aproximado à string pesquisada
- mostrar resultados mais próximos primeiro (ver #39)
- resolver problemas relacionados ao índice da disciplina (e.g. Cálculo I, II, III, IV agora são encontrados corretamente)

_OBS: Um snapshot mostrou o uso de 27MB pela lista de trigramas do banco de dados. Isso equivale a 45% de toda a memória usada pelo matrUSP. Não consegui identificar se esse uso é causado por um leak ou se o algoritmo realmente precisa de tanta memória para funcionar. Vou tentar avaliar melhor._ 

_OBS2: não sabia se aqui era pra escrever em inglês ou não_

Samlibardi committed:
approx-search implemented

implemented search algorithm based on trigram-search

Samlibardi committed:
Better weighting of trigrams

Trigrams now weighted based on global and local (each lecture) frequency

Samlibardi committed:
Fixed whitespace error

pedropaulovc commented:
Obrigado pelo pull request! Revisarei nos próximos 2 dias. Primeiro preciso entender como a busca por trigramas funciona.

Samlibardi commented:
Acredito que seja possível fazer pequenos ajustes para melhorar a precisão das pesquisas. Estarei submetendo as mudanças que fizer.

Samlibardi committed:
better acronym search

by removing articles and conjunctions and adding special trigrams for sequential first letters, it is now posible to do searchs: "GA" -> "Geometria Analitica", "SD" -> "Sistemas Digitais", "EDO" -> "Equacoes Diferenciais Ordinarias"

Samlibardi committed:
support search by arabic numerals

Now properly parses arabics as romans at the end of the search string (2 becomes II, 9 becomes IX etc) in order to be able to search for Calculo 2, Fisica 4, Canto Coral 10 etc.

Samlibardi committed:
Quick fix for a920ee3

Added a few missing cases and excluded if it is the first word

Samlibardi committed:
More cases for a920ee3

pedropaulovc commented on js/database.js:
```diff
@@ -17,71 +17,113 @@ function Database() {
-
-		return tmp.length * value;
+	function romanize(num) {
+	  var lookup = {M:1000,CM:900,D:500,CD:400,C:100,XC:90,L:50,XL:40,X:10,IX:9,V:5,IV:4,I:1},roman = '';
```
Por favor normalize a indentação. Algumas linhas usam 2 espaços e outras usam tab.

pedropaulovc commented on js/database.js:
```diff
@@ -17,71 +17,113 @@ function Database() {
+					words.splice(i,1);
+					i--;
+					continue;
 					break;
```
Este `break` ainda é necessário?

pedropaulovc commented on js/database.js:
```diff
@@ -17,71 +17,113 @@ function Database() {
-				searchWholeString[i].lastIndex = 0;
-				if(searchWholeString[i].test(haystack.code)) {
-					exactly = true;
+	function trigramsFromString(str) {
```
Eu apliquei as suas mudanças no [ambiente de homologação](http://bcc.ime.usp.br/matrusp_homolog/) do MatrUSP e fiz alguns testes:

* Disciplinas com numerais arábicos e romanos. Funcionando muito bem.
* "Calc" / "Cálc": Retorna `MNE0614 - Coma` como o primeiro resultado, o que é estranho. Os outros resultados estão condizentes. 
* "MAC": Também retorna `MNE0614 - Coma` como o primeiro resultado, parece que o MatrUSP está assombrado. 😱
* "EAC": Retorna `EFE0148 - Remo` como o segundo resultado, não parece relacionado com os outros resultados.
 
Você poderia verificar a sua implementação por algum *off-by-one error*? No geral os resultados gerados são muito úteis.

Samlibardi replied:
Acho que isso deve se dar pela taxação do score baseado no tamanho do nome da disciplina. A função que usei foi ~1/log(tamanho). Vou tentar procurar uma função que não dê tanto valor a disciplinas com 3 ou 4 letras apenas (Cor, Coma, Remo etc)

pedropaulovc commented on js/database.js:
```diff
@@ -17,71 +17,113 @@ function Database() {
+			if(i > 0)
+			switch(words[i])
+			{
+				case "DE":
```
Considere converter esse `switch-case` um array/objeto `stopwords_ptBR`. Daí você pode filtrar `words`. Isso ajuda a legibilidade e internacionalização.

pedropaulovc commented on js/database.js:
```diff
@@ -17,71 +17,113 @@ function Database() {
+				case "DO":
+				case "DAS":
+				case "DOS":
+				case "A":
```
Incluir "para" e "por"?

pedropaulovc reviewed (changes requested):
No geral as mudanças estão bem encaminhadas. Por favor dê uma olhada nos meus comentários e assim que resolvidos eu farei o deploy no site oficial. Abraços!

Samlibardi committed:
Fixed indentation and line-breaks

Samlibardi committed:
Improved stopword filtering readability

Samlibardi committed:
Improved trigram extraction

Added parameter to choose trying to parse str as acronym, removed first letter from acronyms (now only sequential first letters).
Lecture names are not parsed as acronyms
Balanced lecture name weighting

Samlibardi commented:
Fiz algumas alterações na maneira como os trigramas são extraídos das palavras, agora apenas a palavra buscada é tratada como possível sigla, e letras no meio da sigla não são tratadas como letras iniciais. Acredito que o problema das matérias "Coma" e "Remo' esteja resolvido.

Samlibardi commented:
Percebi que algumas disciplinas foram colocadas com os algarismos romanos separados por espaço (Ex: Física para engenharia elétrica I I I, pesquisa científica em medicina X V I I I) e algumas tem numeros romanos que não estão no final no nome (Processamento de materiais V I I: Soldagem). Vou fazer alterações para processar esses nomes de forma correta

Samlibardi committed:
Remove whitespace between roman numerals

Samlibardi committed:
Removed restriction on acronym parsing

Not only first words will be treated as acronyms -> can search "lab sd" or "lab icc"

pedropaulovc commented:
Olá Samuel,

Eu não tenho certeza se este PR está relacionado a um comportamento estranho no ambiente de homologação. Eu gostaria que você validasse se você consegue reproduzir:

1. Usando Firefox ou Edge em modo incognito, abra o ambiente de homologação
2. Digite 'a' na caixa de busca de disciplinas
3. Clique no primeiro resultado
4. Digite 'b' na caixa de busca de disciplinas
5. Clique no primeiro resultado
6. Digite 'c' na caixa de busca de disciplinas
7. Clique no primeiro resultado
8. Digite 'd' na caixa de busca de disciplinas
9. Clique no primeiro resultado

Esperado: O MatrUSP deveria listar 4 disciplinas no plano.

Real: Alguns cliques não são registrados. Apenas 2 disciplinas são inseridas no plano.

Samlibardi commented:
Olá. Consegui reproduzir no ambiente de homologação. O erro acontece de fato. No entanto, no meu ambiente local o erro não acontece. O ambiente para esse PR contém exatamente o conteúdo mais recente do branch, sem nenhuma outra modificação. Acredito que o erro deve ser causado por outro código presente no ambiente de homologação que não está atualizado para os commits mais recentes.

pedropaulovc commented:
Fiz um redeploy no ambiente de homologação e realmente o problema se foi. :-)
